### PR TITLE
server: passkey challenge /  self-extend with context shift demo

### PR DESCRIPTION
**Pull request by phymbert**
### Motivation

llama.cpp self-extend logic is reimplementing, and  it [has been discussed](https://github.com/ggerganov/llama.cpp/pull/5710#discussion_r1501885613) to add a [passkey](https://github.com/ggerganov/llama.cpp/blob/master/examples/passkey/README.md) scenario in the server test framework:

- #5104
- #5691
- #4810

### Changes

1. Add model metadata field in the `/v1/models` endpoint and test it. This is useful to get back the `n_ctx_train` or `n_embd` for example. Also, it allows to assert test scenario. [Models available](https://github.com/ggerganov/llama.cpp/compare/tests/server/passkey?expand=1#diff-18972577a8f194e5249a95052e321a055ee5d252e6823ff01fe9d5620f90125dR87-R91) scenario added.
2. Download models from HF using the already included [huggingface_hub](https://huggingface.co/docs/hub/models-downloading) python dependency
3. Support of `--grp-attn-n` and `--grp-attn-w` in tests
4. Add a passkey feature which generates a long prompt of junk and inserts a secret passkey number inside it. The prompt size exceeds the model n_ctx_train and self-extend context shift is triggered.
5. Do not truncate the prompt anymore if the group attention factor is set

### Tested models
`Device 0: NVIDIA GeForce RTX 3050 Laptop GPU, compute capability 8.6, VMM: yes`

- [Phi 2](https://huggingface.co/TheBloke/phi-2-GGUF):  `n_ctx_train=2048, n_ctx=8192, n_prompt=6064, n_batch=512, n_ga=16, n_ga_w=512`
- [Llama-2-7B](https://huggingface.co/TheBloke/Llama-2-7B-GGUF):  `n_ctx_train=4096, n_ctx=16384, n_prompt=12064, n_batch=512, n_ga=4, n_ga_w=512`
- [Mixtral-8x7B-v0.1](https://huggingface.co/TheBloke/Mixtral-8x7B-v0.1-GGUF):  `n_ctx_train=4096(*8?), n_ctx=16384, n_prompt=12064, n_batch=512, n_ga=4, n_ga_w=512`

### How to test:

```shell
N_GPU_LAYERS=5 LLAMA_SERVER_BIN_PATH=../../../bin/server DEBUG=ON ./tests.sh --tags passkey --no-capture --no-skipped --stop
```
You can change examples in [passkey.feature](https://github.com/ggerganov/llama.cpp/compare/tests/server/passkey?expand=1#diff-e165468bd92dedf7e1ae733c0db37855ce7824014cd3015315e5319c78ca4803R49-R51)

### Issues
- [ ]  prompt processing and self-extend are blocking the server main thread loop. We need to refactor thread usage per slot.

### CODE REVIEW

File: examples/server/server.cpp [hunk 1]
```diff
@@ -1709,8 +1709,8 @@ struct llama_server_context
                     }
                     slot.params.n_keep = std::min(slot.n_ctx - 4, slot.params.n_keep);
 
-                    // if input prompt is too big, truncate it
-                    if (slot.n_prompt_tokens >= slot.n_ctx)
+                    // if input prompt is too big, truncate it, if group attention self-extend is disabled
```
Comment (phymbert) [inline examples/server/server.cpp:1712]: @ggerganov Please confirm my understanding ?

**Comment by phymbert** [comment]
Hold on, I used `input_prefix` and `input_suffix` but they are used only for `/infil` ... fixing

**Comment by phymbert** [comment]
@Maximilian-Winter FYI, appreciate your feedback

File: examples/server/tests/features/passkey.feature [hunk 2]
```diff
@@ -46,5 +46,8 @@ Feature: Passkey / Self-extend with context shift
     Then <n_predicted> tokens are predicted matching <re_content>
 
     Examples:
-      | hf_repo             | hf_file           | n_ctx_train | ngl | n_ctx | n_batch | n_ga | n_ga_w | n_junk | i_pos | passkey | n_predicted | re_content |
-      | TheBloke/phi-2-GGUF | phi-2.Q4_K_M.gguf | 2048        | 5   | 16384 | 512     | 16   | 512    | 250    | 50    | 42      | 1           | 42         |
+      | hf_repo                         | hf_file                     | n_ctx_train | ngl | n_ctx | n_batch | n_ga | n_ga_w | n_junk | i_pos | passkey | n_predicted | re_content |
+      | TheBloke/phi-2-GGUF             | phi-2.Q4_K_M.gguf           | 2048        | 5   | 8192  | 512     | 16   | 512    | 250    | 50    | 42      | 1           | 42         |
+      | TheBloke/Llama-2-7B-GGUF        | llama-2-7b.Q2_K.gguf        | 4096        | 3   | 16384 | 512     | 4    | 512    | 500    | 300   | 1234    | 5           | 1234       |
+      | TheBloke/Mixtral-8x7B-v0.1-GGUF | mixtral-8x7b-v0.1.Q2_K.gguf | 4096        | 2   | 16384 | 512     | 4    | 512    | 500    | 100   | 0987    | 5           | 0987       |
+
```
Comment (ggerganov) [inline examples/server/tests/features/passkey.feature:53]: I would leave only the `phi` test and change `n_ga = 4`.

Is it possible to add a "negative" test which uses `n_ga = 2` and checks that it fails to match the passkey?

File: examples/server/server.cpp (hunk 1 above)
Comment (ggerganov) [reply examples/server/server.cpp:1712]: Yes, this is correct

**Review by ggerganov — APPROVED** [review]

**Pull request merged**